### PR TITLE
Rewards Pallet

### LINKED ISSUE #594: Parachain Rewards Pallet
### Introduction
Polkadex secured a para chain slot on Polkadot with 973k DOT ($13.8M) contributed to the Polkadex Crowdloan. All the participants who loaned their DOT tokens will receive PDEX tokens. 25% of PDEX tokens receive as rewards can be claimed by participants immediately while the rest of the 75% will linearly unlock. The treasury account will be a pallet account from which rewards funds will be transferred to participants. 

### Flow Diagram
reward_flow

### Storage Required

**InitializeRewards Map**
Key: reward_id
value: (start_block, end_block, initial_percentage)

**Distributor Map**
Key 1: reward_id
Key 2: account_id
value: (total_reward_amount, claim_amount, is_initial_rewards_claimed, is_initialized, lock_id, last_block_rewards_claim)

### Extrinsic Required

1) `create_reward_cycle`

The extrinsic will be used to start a new reward cycle and can only be called by governance. 
`Parameters`
origin: The sudo/governance address
start_block: The block from where we will start releasing rewards
end_block: By the end_block all rewards should have been released. 
reward_id: The id for which vesting needs to be set. 

----------------------------------------------------

2) `initialize_claim_rewards`
The extrinsic will transfer and lock users' rewards in the user account. It will get user reward_info from the constant hashmap. 
`Parameters`
origin: The users address
reward_id: The id for which vesting needs to be set. 

----------------------------------------------------

3) `claim`
The user will use the extrinsic to claim rewards
`Parameters`
origin: The users address
reward_id: The id for which vesting needs to be set. 

----------------------------------------------------

ToDos
- [x] Design Extrinsics
- [x] Design mocking environment for test cases
- [x] Test Cases

### LINKED ISSUE #603: BenchMarks for Rewards Pallet
ToDos to close the issue 
-  [x] new weight annotations using outputs from actual benchmarking

### PR DESCRIPTION
**Pull request by nakul1010**
Closes #594.

### CODE REVIEW

File: pallets/rewards/src/lib.rs
Comment (nakul1010) [inline pallets/rewards/src/lib.rs:23]: Add the crowd loan reward information as a doc in the pallet.

File: pallets/rewards/src/lib.rs
Comment (nakul1010) [inline pallets/rewards/src/lib.rs]: Remove this check as the crowd loan start block will be an already mined block and not an upcoming block.

File: pallets/rewards/src/lib.rs
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Is it required to allow this on the global module level (each child module will allow this as well)? I assume this is an exception so I'd recommend to add attribute (without `!`) only in proper place e.g. line # 201

File: pallets/rewards/src/lib.rs
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Same is here, please move to the line # 9 of `crowdloan_rewardees.rs` module.

File: pallets/rewards/src/lib.rs
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Doc second arg (`id`) is different from declared ('reward_id') please make them update name in one of places.

File: pallets/rewards/src/lib.rs
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs:204]: Else case should be handled here I think, at least because on the line above you are taking value from the storage. Plus the line # 231 asserted basically if this `if` worked. So I assume in else case could be returned `UserNotEligible` error? is it user not eligible or reward is to low to be claimed, mby some proper error would be more preferable in this situation?

File: pallets/rewards/src/lib.rs
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: I see regression in tests comparing to the previous PR. [This](https://github.com/Polkadex-Substrate/Polkadex/blob/c4cf4a4ce9175bd3b76c0c658b02a789f6da237b/pallets/rewards/src/tests.rs#L778-L782) case is not covered here. Please add such assertion back.

File: runtime/Cargo.toml
Comment (serhii-temchenko) [inline runtime/Cargo.toml]: Please add space after `{`. Same on line 105 and 111.

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

File: pallets/rewards/src/lib.rs
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: Added a similar test case in `initialize_claim_rewards`, which helped me to find a bug, thanks.

**Review by serhii-temchenko — DISMISSED** [review]

File: pallets/rewards/Cargo.toml
Comment (35359595) [inline pallets/rewards/Cargo.toml]: pallet name should start with `pallet-` prefix

File: pallets/rewards/Cargo.toml
Comment (35359595) [inline pallets/rewards/Cargo.toml]: why import benchmarking if there are no benchmarks implemented?

File: pallets/rewards/Cargo.toml
Comment (35359595) [inline pallets/rewards/Cargo.toml]: why import benchmarking if there are no benchmarks implemented?

File: pallets/rewards/src/crowdloan_rewardees.rs
Comment (35359595) [inline pallets/rewards/src/crowdloan_rewardees.rs:8]: is this done? why PR is ready if this TODO is still in place?

File: pallets/rewards/src/lib.rs
Comment (35359595) [inline pallets/rewards/src/lib.rs]: change to `-2023`

File: pallets/rewards/src/lib.rs
Comment (35359595) [inline pallets/rewards/src/lib.rs]: No weights through benchmarks

File: pallets/rewards/src/lib.rs
Comment (35359595) [inline pallets/rewards/src/lib.rs]: No weights through benchmarks

File: pallets/rewards/src/lib.rs
Comment (35359595) [inline pallets/rewards/src/lib.rs]: No weights through benchmarks

**Review by 35359595 — CHANGES_REQUESTED** [review]
No benchmarks implemented - please implement

File: pallets/rewards/Cargo.toml
Comment (nakul1010) [reply pallets/rewards/Cargo.toml]: removed

File: pallets/rewards/Cargo.toml
Comment (nakul1010) [reply pallets/rewards/Cargo.toml]: removed

File: pallets/rewards/src/lib.rs
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: Issue  #629 opened for it.

File: pallets/rewards/src/lib.rs
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: Issue https://github.com/Polkadex-Substrate/Polkadex/issues/629 opened for it.

File: pallets/rewards/src/lib.rs
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: Issue https://github.com/Polkadex-Substrate/Polkadex/issues/629 opened for it.

File: pallets/rewards/src/crowdloan_rewardees.rs
Comment (nakul1010) [reply pallets/rewards/src/crowdloan_rewardees.rs:8]: We are still waiting for the response from the crowd loan partners to update the map. When the issue will be solved the hashmap will be updated.

Comment (35359595) [reply pallets/rewards/src/crowdloan_rewardees.rs:8]: so should we wait with review / merge until that's done?

File: pallets/rewards/src/lib.rs
Comment (35359595) [reply pallets/rewards/src/lib.rs]: can we wait with merging this PR until benchmarks and weights are merged here?

**Comment by nakul1010** [comment]
Reward Calculation [PR](https://github.com/Polkadex-Substrate/Reward-Calculation/pull/3) needs to be merged to update the hashmap.

File: pallets/rewards/src/crowdloan_rewardees.rs
Comment (nakul1010) [reply pallets/rewards/src/crowdloan_rewardees.rs:8]: merge until that's done

File: pallets/rewards/Cargo.toml
Comment (serhii-temchenko) [inline pallets/rewards/Cargo.toml]: Please add empty line at the end of file.

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]
> Reward Calculation [PR](https://github.com/Polkadex-Substrate/Reward-Calculation/pull/3) needs to be merged to update the hashmap.

Approved. Please resolve conflict in this PR and fix minor code style issue from the comment below.

File: pallets/rewards/src/lib.rs
Comment (35359595) [inline pallets/rewards/src/lib.rs]: This ensure is dublicated by next if let reading same value twice. remove

File: pallets/rewards/src/lib.rs
Comment (35359595) [inline pallets/rewards/src/lib.rs]: this logic is shady at best. you did check on line 196, then inserted, then check again and then mutate what you JUST inserted. 2 reads and 2 writes for no reason at all. rework into proper logic

File: pallets/rewards/src/lib.rs
Comment (35359595) [inline pallets/rewards/src/lib.rs]: remove these two checks and return errors instead of "sanity check"s below to reduce reads by 2

File: pallets/rewards/src/lib.rs
Comment (35359595) [inline pallets/rewards/src/lib.rs:343]: @Gauthamastro is there a document for this logic? Otherwise, please verify this flow is correct.

**Review by 35359595 — CHANGES_REQUESTED** [review]

File: pallets/rewards/src/crowdloan_rewardees.rs
Comment (35359595) [reply pallets/rewards/src/crowdloan_rewardees.rs:8]: @Gauthamastro to confirm before I can resolve without change

File: pallets/rewards/src/lib.rs
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: Done

File: pallets/rewards/src/lib.rs
Comment (35359595) [inline pallets/rewards/src/lib.rs]: `transfer_pdex_rewards()` already returns `DispathResult` with proper reason. What's the point of overwriting it here with generic `TransferFailed` error?

**Review by 35359595 — CHANGES_REQUESTED** [review]

File: pallets/rewards/src/lib.rs
Comment (nakul1010) [reply pallets/rewards/src/lib.rs]: Removed ensured and passed back the error to the calling function.

File: pallets/rewards/src/crowdloan_rewardees.rs
Comment (nakul1010) [reply pallets/rewards/src/crowdloan_rewardees.rs:8]: Added the hashmap and to generate the hashmap [Reward Calculation ](https://github.com/Polkadex-Substrate/Reward-Calculation) repo was used.

**Review by 35359595 — DISMISSED** [review]

File: pallets/rewards/Cargo.toml
Comment (serhii-temchenko) [reply pallets/rewards/Cargo.toml]: @nakul1010 Still no empty line

File: pallets/rewards/src/lib.rs
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: Why is it here? Please remove.

File: pallets/rewards/src/lib.rs
Comment (serhii-temchenko) [inline pallets/rewards/src/lib.rs]: `TryInto` is not used. Please remove.
```rust
	use sp_std::cmp::min;
```

**Review by serhii-temchenko — CHANGES_REQUESTED** [review]

**Review by serhii-temchenko — DISMISSED** [review]

**Review by 35359595 — DISMISSED** [review]

**Review by zktony — APPROVED** [review]

**Review by serhii-temchenko — APPROVED** [review]

File: pallets/rewards/src/lib.rs
Comment (serhii-temchenko) [reply pallets/rewards/src/lib.rs:343]: > @Gauthamastro is there a document for this logic? Otherwise, please verify this flow is correct.

Created issue for this. #657.

File: pallets/rewards/src/crowdloan_rewardees.rs
Comment (serhii-temchenko) [reply pallets/rewards/src/crowdloan_rewardees.rs:8]: > @Gauthamastro to confirm before I can resolve without change

Created issue for this. #657.

**Pull request merged**
